pkg_summary-utils/files/pkg_grep_summary.1
.\"	$NetBSD: pkg_grep_summary.1,v 1.29 2011/11/09 14:18:05 cheusov Exp $
.\"
.\" Copyright (c) 2008-2010 by Aleksey Cheusov (vle@gmx.net)
.\" Absolutely no warranty.
.\"
.\" ------------------------------------------------------------------
.de VB \" Verbatim Begin
.ft CW
.nf
.ne \\$1
..
.de VE \" Verbatim End
.ft R
.fi
..
.\" ------------------------------------------------------------------
.TH PKG_GREP_SUMMARY 1 "Jan 29, 2008" "" ""
.SH NAME
pkg_grep_summary \- tool for grepping pkgsrc summary file
.SH SYNOPSIS
.BI "pkg_grep_summary -h"
.br
.BI "pkg_grep_summary [option]" " field awk_expression"
.br
.BI "pkg_grep_summary [option] -e " field
.br
.BI "pkg_grep_summary [option] -m" " field regexp"
.br
.BI "pkg_grep_summary [option] -s" " field string"
.br
.BI "pkg_grep_summary [option] -S" " field filename"
.br
.BI "pkg_grep_summary [option] -t" " strategy field condition"
.br
.BI "pkg_grep_summary -T"
.br
.SH DESCRIPTION
.B pkg_grep_summary
takes package summaries on stdin and
outputs those matching the specified condition.
Format of summary is described in
.BR pkg_summary(5) .
Fields are PKGPATH, PKGNAME, DEPENDS, MAINTAINER
and so on, actually anything separated by 
.B =
sign.
In addition to fields really present on input a few artificial fields
are supported: PKGBASE, PKGPATHe, PKGPAIR, PKGPABA and PKGPANA. PKGBASE
is a PKGNAME with package version stripped. PKGPATHe (enriched PKGPATH)
is PKGPATH with ASSIGNMENTS
(if present) separated by colon symbol, for example,
.IR lang/py-psyco:PYTHON_VERSION_REQD=25 .
PKGPAIR is a PKGPATH and PKGBASE separated by comma character, for example,
.IR net/dhisd,dhisd .
PKGPABA is a synonym for PKGPAIR. PKGPANA is
a PKGPATH and PKGNAME separated by
comma, for example textproc/dict-server,dict-server-1.11.2.
.B pkg_grep_summary
is a search tool somewhat similar to
.BR grep(1) ", " PKGSRCDIR/pkglocate ", " pkgfind
and others. Unlike 
.B grep(1)
it uses summary information about one package instead of line of text
as an indivisible block of text.  Unlike
.BR pkgfind " and " PKGSRCDIR/pkglocate
it doesn't scan pkgsrc source tree.
For scanning pkgsrc source tree you may use
.BR pkg_src_summary(1) " or " pkg_micro_src_summary(1) .
The following variables can be used in
.IR awk_expression ": "
"fname"
(field name to check),
"fvalue"
(part of line after = sign) and associative array
"fields"
mapping field names to values which is available only if 
.I field
is `.' (single dot).
.SH OPTIONS
.TP
.B "-h"
display help message.
.TP
.B "-i"
ignore case. Before matching
.I field
is converted to lower case.
.TP
.B "-v"
invert the sense of match
.TP
.B "-r"
exit status is non-zero if no matches were found.
.TP
.B "-R"
option
.B -r
is implied and "No matches found" is output to stderr.
.TP
.BI "-f" " fields"
a list of fields to output, by default the whole package summary is output.
.TP
.BI "-e" " field"
outputs summaries in which
.I "field"
is either empty or absent.
.TP
.BI "-m" " field regexp"
outputs summary if
.I "regexp"
matches
.I "field"
.TP
.BI "-s" " field string"
outputs summary if
.I "field"
is equal to
.I "string"
.TP
.BI "-S" " field filename"
reads strings from
.IR filename ,
one string per line, and outputs summary given on stdin if
.I "field"
is equal to one of read strings.
.TP
.BI "-t" " strategy field condition"
outputs summary if
.I condition
matches
.I field
according to search strategy
.IR strategy .
Available search strategies are:
.IR exact ", " prefix ", " suffix ", " substring ", " word ", "
.IR first ", " last ", " re ", " strfile ", " strlist ", " awk ", " kw
.RI " (" keywords ") and " empty .
Search strategy 
.I re
is for matching using regexp (ERE).
Search strategies
.IR first ", " last " and " word
are for matching first, last or any separate word within 
.IR field .
With
.IR kw " (" keywords ")"
.I condition
is treated as a list of words each of which should present in
.IR field .
With
search strategy
.IR strlist ", " condition
is treated as a list of words at least one of which should present in
.IR field .
.TP
.B "-T"
display all search strategies available for
.B -t
and exits. Doubled
.I -T
will show search strategies with their descriptions separated by TAB character.
.SH NOTES
.IR "-t exact <field> <string>" " and " "-s <field> <string>" " are equivalents"
.br
.IR "-t re <field> <regexp>" " and " "-m <field> <regexp>" " are equivalents"
.br
.IR "-t strfile <field> <filename>" " and " "-S <field> <filename>" " are equivalents"
.br
.IR "-t awk <field> <awk_expr>" " and " "<field> <awk_expr>" " are equivalents"
.br
.IR "-t empty <field> ''" " and " "-e <field>" " are equivalents"
.SH EXAMPLES
.VB
pkg_grep_summary -s PKGPATH 'devel/libjudy'

pkg_grep_summary -m PKGPATH '^wip/'

pkg_info -Xa | pkg_grep_summary SIZE_PKG 'fvalue+0 > 100000000'

pkg_grep_summary PKGNAME 'fvalue ~ /judy/' \\
      < /usr/pkgsrc/packages/pkg_summary.txt

pkg_grep_summary PKGPATH 'fvalue ~ /^lang\\//' \\
      < /usr/pkgsrc/packages/pkg_summary.txt

pkg_grep_summary DEPENDS 'fvalue ~ /libX11/' \\
      < /usr/pkgsrc/pkg_src_summary.txt

pkg_grep_summary COMMENT 'toupper(fvalue) ~ /DNS/' \\
      < /usr/pkgsrc/pkg_src_summary.txt

pkg_info -QPKGPATH -a > /tmp/wanted_pkgpaths.txt
pkg_grep_summary -S PKGPATH /tmp/wanted_pkgpaths.txt \\
      < /usr/pkgsrc/packages/pkg_summary.txt

pkg_grep_summary -t word COMMENT DNS \\
      < /usr/pkgsrc/pkg_src_summary.txt

pkg_bin_summary -f PKGNAME,automatic,REQUIREDBY |
pkg_grep_summary . 'fields ["automatic"] == "yes" &&
                    fields ["REQUIREDBY"] == ""'

pkg_grep_summary -i -t kw COMMENT 'dns server' \\
      < /usr/pkgsrc/packages/pkg_summary.txt

pkg_grep_summary -t strlist MAINAINER 'cheusov@tut.by vle@gmx.net' \\
      < /usr/pkgsrc/packages/pkg_summary.txt
.VE
.SH SEE ALSO
.BR pkg_summary-utils(7),
.BR pkg_summary(5)
.BR pkg_src_summary(1)
.BR pkg_micro_src_summary(1)
.BR pkg_info(1)
.SH AUTHOR
Aleksey Cheusov <vle@gmx.net>


hepmc/ChangeLog
2010-08-18  Lynn Garren

      * src/GenEventStreamIO.cc, StreamHelpers.cc: 
        cleanup unassociated event remnants if corrupt data is found

2010-07-30  Lynn Garren & Andy Buckley

      * src/GenEventStreamIO.cc: create GenCrossSection, PdfInfo, and HeavyIon
        on the stack rather than the heap to fix a memory leak

  --------------------------  HepMC-2.06.01  --------------------------
2010-06-01  Lynn Garren

      * HepMC/GenRanges.h: remove ConstGenVertexParticleRange since 
        there is neither a GenVertex::particle_const_iterator nor a 
        GenVertex::vertex_const_iterator 
      * HepMC/GenVertex.h: remove unimplemented 
        ConstGenVertexParticleRange particles()


  --------------------------  HepMC-2.06.00  --------------------------
  --------------------------  HepMC-2.06.00.b02  --------------------------
2010-05-13  Lynn Garren

      * HepMC/Version.h: version() now takes an optional output stream

      * HepMC/GenEvent.h: add helper method for reading weight names
      * src/GenEventStreamIO.cc: use new helper method to read weight names
      * testHepMCVarious.input: change order of some event info lines
      * test/testIOGenEvent.input: add named weights for testing

2010-05-12  Lynn Garren

      * "make check" now works if momentum units are MEV

2010-05-11  Lynn Garren

      * rename HepMC/GenEventIterators.h to HepMC/GenRanges.h
      
      * src/GenEventStreamIO.cc: fix a bug reading unnamed (old style) weights
      
      * test/testHepMC: add more weight testing

  --------------------------  HepMC-2.06.00.b01  --------------------------

2010-04-05  Lynn Garren

      * HepMC/StreamInfo.h: add reading_event_header flag for use by streaming input
      * src/GenEventStreamIO.cc: Input no longer expects a fixed order of 
        header lines and will ignore lines that are not of a known type.

      * HepMC/HerwigWrapper.h: add hwbeam
      * rename PythiaWrapper6_2.h to PythiaWrapper6_4.h 
        Add structs for more common blocks.

2010-04-02  Lynn Garren

      * HepMC/IteratorRange.h: put the enum in its own header
      * HepMC/GenEvent.h: add vertex_range() and particle_range()
      * HepMC/GenParticle.h: add particles_in(range) and particles_out(range)
      * HepMC/GenVertex.h: add particles(range), 
        particles_in(GenParticle,range), and particles_out(GenParticle,range)
      * test/testHepMCIteration.cc: use the new methods

2010-03-20  Lynn Garren

      * HepMC/GenEventIterators.h: add ConstGenVertexParticleRange,
        ConstGenParticleProductionRange, and ConstGenParticleEndRange
      * HepMC/WeightContainer.h: method to check for weight name is has_key()
      * src/GenEventStreamIO.cc: use quotes to delineate weight names

2010-02-08  Lynn Garren

      * remove unused PythiaWrapper5_720.h, PythiaWrapper6_152.h, and PythiaWrapper6_152_WIN32.h
      * combine HerwigWrapper6_4.h and HerwigWrapper.h in a single file
      * HepMC/PythiaWrapper.h: add getPythiaCrossSection function
      * HepMC/HerwigWrapper.h: add getHerwigCrossSection function
      * fio/PythiaWrapper.cc: implement getPythiaCrossSection function
      * fio/HerwigWrapper.cc: implement getHerwigCrossSection function
      * remove redundant examples/HerwigHelper.h.

2010-02-04  Lynn Garren

      * doc/Makefile.am, examples/Makefile.am: use pkgdatadir to define the instation directory
        The default installation directory is now ${prefix}/share/HepMC/doc
        Override with --datadir=/some/directory/path

      * HepMC/Polarization.h: add is_defined() and set_undefined()
        is_defined returns false if the polarization is undefined
	set_undefined unsets the defined flag and zeroes the Polarization

2010-01-28  Lynn Garren

      * HepMC/GenEventIterators.h: This new header defines a set of classes
        which provide the appropriate iterators ranges.  They can be used to 
	simplify code and also directly with such utilities as Boost foreach.
	These new classes are 
	GenEventVertexRange, ConstGenEventVertexRange, 
	GenEventParticleRange, ConstGenEventParticleRange, 
	GenVertexParticleRange, 
	GenParticleProductionRange, and GenParticleEndRange
      * test/testHepMCIteration.cc: use the new iterator clases and
        also illustrate the use of std::for_each
 
2010-01-26  Lynn Garren

      * test/testHepMCIteration: illustrate use of std::for_each
 
2010-01-25  Lynn Garren

      * HepMC/WeightContainer.h: add map-like functionality, allowing users to
        name weights, while retaining the original vector interface for 
	complete backwards compatibility
	If a name is not specified for a particular weight, the default
	name is the weight's vector index.
	Also add == and != operators.
      * HepMC/GenEvent.h: add a private method used by the stream I/O
      * src/GenEventStreamIO.cc: add a new line for weight names
        The weight name line is optional when reading input and is 
	written only if the event weights object is not empty.
      * test/testWeights.cc: add tests for WeightContainer
      * test/testHepMC: test I/O of named weights
      * examples/testPythiaCopies.cc: set some arbitrary weights
 
2010-01-09  Lynn Garren

      * removed ParticleDataTable.h and all supporting references.

 --------------------------  HepMC-2.05.01  --------------------------

2010-01-04  Lynn Garren

      * HepMC/GenEvent.h: add write_cross_section(std::ostream&)
      * src/GenEvent.cc: output the cross section (if it exists) when printing an event
      * src/GenVertex.cc, GenParticle.cc: save and restore output stream state
        when calling print methods
	
      * examples: Calculate Pythia and Herwig cross sections using
        getPythiaCrossSection from examples/PythiaHelper.h and 
	getHerwigCrossSection from examples/HerwigHelper.h.
        Set the units explicitly.
      * test/testHepMC.cc.in: call write_cross_section()

2009-12-28  Lynn Garren

       * examples/VectorConversion.h: inline the functions


pkg_summary-utils/files/pkg_grep_summary.awk
# Copyright (c) 2010, Aleksey Cheusov <vle@gmx.net>
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
#
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#     * Redistributions in binary form must reproduce the above
#       copyright notice, this list of conditions and the following
#       disclaimer in the documentation and/or other materials provided
#       with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#use "has_prefix.awk"
#use "has_suffix.awk"
#use "psu_funcs.awk"

#env "LC_ALL=C"

BEGIN {
	_gs_matched = -1 # -1 - unknown, 0 - false, 1 - true
	_gs_count   = 0

	_sg_multiline ["PLIST"]       = 1
	_sg_multiline ["DESCRIPTION"] = 1
	_sg_multiline ["DEPENDS"]     = 1
	_sg_multiline ["REQUIRES"]    = 1
	_sg_multiline ["PROVIDES"]    = 1
	_sg_multiline ["CONFLICTS"]   = 1

	keep_fields = 0

	invert += 0

	# required fields
	gsub(/,/, " ", reqd_fields_)
	reqd_fields_cnt = split(reqd_fields_, f)
	for (i=1; i <= reqd_fields_cnt; ++i){
		reqd_fields [f [i]] = 1
	}
	delete f
}

function match_first_word (s, word){
	if (s == word)
		return 1
	else if (!has_prefix(s, word))
		return 0
	else{
		return substr(s, length(word)+1, 1) ~ /^[^A-Za-z0-9]$/
	}
}

function match_last_word (s, word){
	if (s == word)
		return 1
	else if (!has_suffix(s, word))
		return 0
	else
		return substr(s, length(s)-length(word), 1) ~ /^[^A-Za-z0-9]$/
}

function match_word (s, word,                  idx){
	if (s == word)
		return 1

	idx = index(s, word)
	if (!idx)
		return 0

	if (idx > 1 && substr(s, idx-1, 1) ~ /[A-Za-z0-9]$/)
		return 0

	idx += length(word)
	if (idx <= length(s) && substr(s, idx, 1) ~ /[A-Za-z0-9]$/)
		return 0

	return 1
}

function match_keywords (s,         cnt_s, arr_s, set_s, i){
    if (s !~ re_kw)
		return 0

	cnt_s = split(s, arr_s, /[^A-Za-z0-9]/)
	for (i=1; i <= cnt_s; ++i){
		set_s [arr_s [i]] = 1
	}

	for (i=1; i <= cnt_kw; ++i){
		if (! (arr_kw [i] in set_s)){
			return 0
		}
	}

	return 1
}

function update_skip (){
	if (ic)
		fvalue = tolower(fvalue)

	_gs_matched = grep_summary__condition()

	if (_gs_matched == 0 && (fname in _sg_multiline)){
		_gs_matched = -1
	}

	if (invert && _gs_matched >= 0){
		_gs_matched = 1-_gs_matched;
	}

	if (_gs_matched == 1){
		for (i=0; i < _gs_count; ++i)
			print _gs_accu [i]

		delete _gs_accu
		_gs_count = 0
	}
}

_gs_matched == 0 && NF > 0 {
	next
}

{
	pos = index($0, "=")
	if (pos > 0){
		fname  = substr($0, 1, pos-1)
		fvalue = substr($0, pos+1)
	}else{
		fname  = ""
		fvalue = ""
	}
	fname_orig = fname
}

function check_PKGPATHe (){
	if (_gs_assigns != "" && _gs_pkgpath != ""){
		fvalue = _gs_pkgpath ":" _gs_assigns
		fname  = "PKGPATHe"
		update_skip()
	}else if (index(_gs_pkgpath, ":") > 0){
		fvalue = _gs_pkgpath
		fname  = "PKGPATHe"
		update_skip()
	}
}

_gs_matched == -1 {
	if (grep_summary__field == "PKGBASE"){
		if (fname == "PKGNAME"){
			fname = "PKGBASE"
			fvalue = pkgname2pkgbase(fvalue)
			update_skip()
		}
	}else if (grep_summary__field == "PKGPATHe"){
		if (fname == "ASSIGNMENTS") {
			_gs_assigns = fvalue
			check_PKGPATHe()
		}else if (fname == "PKGPATH") {
			_gs_pkgpath = fvalue
			check_PKGPATHe()
		}
	}else if (grep_summary__field == "PKGPATH"){
		idx = index(fvalue, ":")
		if (idx > 0)
			fvalue = substr(fvalue, 1, idx-1)
	}else if (grep_summary__field == "PKGPAIR"){
		if (fname == "PKGPATH") {
			_gs_pkgpath = fvalue
		}else if (fname == "PKGNAME") {
			_gs_pkgbase = fvalue
			sub(/-[^-]*$/, "", _gs_pkgbase)
		}

		if (_gs_pkgbase != "" && _gs_pkgpath != ""){
			fvalue = _gs_pkgpath "," _gs_pkgbase
			fname  = "PKGPAIR"
			update_skip()
		}
	}

	if (fname == grep_summary__field || "" == grep_summary__field) {
		update_skip()
	}
}

_gs_matched == 1 && NF > 0 {
	if (!reqd_fields_cnt || (fname_orig in reqd_fields))
		print $0

	next
}

keep_fields {
	fields [fname] = fvalue
}

_gs_matched == -1 && NF > 0 {
	if (!reqd_fields_cnt || (fname_orig in reqd_fields))
		_gs_accu [_gs_count++] = $0
}

NF == 0 {
	if (_gs_matched == -1 && grep_summary__field == "PKGPATHe"){
		fvalue = _gs_pkgpath
		update_skip()
		fvalue = ""
	}
	if (_gs_matched == -1){
		update_skip()
	}
	if (_gs_matched == 1){
		++_gs_matches_cnt
		print ""
	}

	delete _gs_accu
	delete fields
	_gs_count = 0
	_gs_matched = -1

	_gs_assigns = _gs_pkgpath = _gs_pkgbase = ""
}

END {
	if (error && !_gs_matches_cnt){
		if (error_msg)
			print error_msg > "/dev/stderr"

		exitnow(1)
	}
}
